src/NewtonKrylov.fypp

```diff
@@ -0,0 +1,418 @@
+      class(abstract_vector_${type[0]}$${kind}$),                         intent(inout) :: X
+      !! Initial guess for the fixed point, will be overwritten with solution
+      integer,                                            intent(out)   :: info
+      !! Information flag
+      real(${kind}$),                                 optional, intent(in)    :: tolerance
+      real(${kind}$)                                                          :: target_tol
+      !! Target absolute solver tolerance
+      type(newton_${kind}$_opts),                     optional, intent(in)    :: options
+      type(newton_${kind}$_opts)                                              :: opts
+      !! Options for the Newton-Krylov iteration
+      procedure(abstract_linear_solver_${type[0]}$${kind}$),    optional                :: linear_solver
+      !! Linear solver to be used to find Newton step
+      class(abstract_opts),                     optional, intent(in)    :: linear_solver_options
+      class(abstract_opts), allocatable                                 :: solver_opts
+      !! Options for the linear solver
+      class(abstract_precond_${type[0]}$${kind}$),              optional, intent(in)    :: preconditioner
+      class(abstract_precond_${type[0]}$${kind}$), allocatable                          :: precond
+      !! Preconditioner for the linear solver
+      procedure(abstract_scheduler_${kind}$),         optional                :: scheduler
+
+      !--------------------------------------
+      !-----     Internal variables     -----
+      !--------------------------------------
+      
+      class(abstract_vector_${type[0]}$${kind}$), allocatable :: residual, increment
+      real(${kind}$) :: rnorm, tol
+      logical :: converged, has_precond, has_solver_opts
+      integer :: i, maxiter, maxstep_bisection
+      character(len=256) :: msg
+      procedure(abstract_linear_solver_${type[0]}$${kind}$), pointer :: solver => null()
+      procedure(abstract_scheduler_${kind}$), pointer :: tolerance_scheduler => null()
+
+      ! Newton-solver tolerance
+      if (present(tolerance)) then
+         target_tol = tolerance
+      else
+         target_tol = atol_${kind}$
+      end if
+      ! Newton-Krylov options
+      if (present(options)) then
+         opts = options
+      else
+         opts = newton_${kind}$_opts()
+      end if
+      ! Linear solver
+      if (present(linear_solver)) then
+         solver => linear_solver
+      else
+         solver => gmres_${type[0]}$${kind}$
+      end if
+      ! Linear solver options ?
+      if (present(linear_solver_options)) then
+         has_solver_opts = .true.
+         allocate(solver_opts, source=linear_solver_options)
+      else
+         has_solver_opts = .false.
+      end if
+      ! Preconditioner ?
+      if (present(preconditioner)) then
+         has_precond = .true.
+         allocate(precond, source=preconditioner)
+      else
+         has_precond = .false.
+      end if
```

Comment:
I'm not sure these are necessary. See my next comment regarding the linear solver.